src/Npgsql/PoolManager.cs

```diff
@@ -24,11 +24,12 @@ static class PoolManager
         internal static bool TryGetValue(string key, [NotNullWhenTrue] out ConnectorPool? pool)
         {
             // Note that pools never get removed. _pools is strictly append-only.
+            var nextSlot = Interlocked.CompareExchange(ref _nextSlot, 0, 0);
```

Comment:
No, no, no. There is `Volatile.Read` for such cases (: Probably the same thing should be done for `_pools` too, otherwise, it could be read a bit later than it should. Or it could be better to mark the fields as `volatile`, so the compiler will use `Volitile.Read` and `Volatile.Write` automatically.

@roji Review the logic here please, because in some cases the code looks dangerous. For example, `ClearAll` doesn't introduce a lock, but it can be invoked from user code.

Reply:
Agree with @YohDeadfall - please make both `_nextSlot` and `_pools` volatile fields.

For `ClearAll()` we should just introduce a lock and be done with it. This is a very rarely-used method and we shouldn't use risky lock-free logic there.

Reply:
Thanks guys, I've updated with the latest suggestions. Let me know if there's anything else.